Chapter4 (#50)
UBC-MDS/ds-toolbox: chapters/en/module4.md

---
title: 'Module 4: Getting groovy with Git and GitHub'
description:
  'View your git history, travel back in time, deal with merge conflicts and other useful tools' 
prev: ../../module3
next: ../../module5
type: chapter
id: 4
---

> <exercise id="1" title="Viewing your Git history" type="slides,video">

Reviewer: I think for slide deck titles, we should use either the -ing form of the verbs or their simple form, e.g. either "Viewing, Restoring, Reverting, Dealing, etc." or "View, Restore, Revert, Deal, etc.". I personally prefer the latter.

<slides source="module4/module4_01" shot="0" start="0:002" end="3:40">
</slides>

</exercise>

<exercise id='2' title="Test Your Knowledge">

Clone the repo https://github.com/UBC-MDS/equine_numbers_value_canada. What is the hash or the first commit of the project? 

<choice>
<opt text='cc400df4b2a382f653abd0d9e514f6daae98d2ad'>
</opt>
<opt text='cc400df'>
</opt>
<opt text='0896801'>
</opt>
<opt text='0896801fdd38d461414e0b061b2b2ef3395f7dd0'>
</opt>
<opt text='A and B are correct'>
</opt>
<opt text='C and D are correct' correct='true'>
</opt>
<opt text='B and C are correct'>
</opt>
</choice>
</exercise>

<exercise id='3' title="Restoring an older version of a file"  type='slides, video'>
<slides source='module4/module4_02' shot='0' start='3:42' end='4:35'> </slides>
</exercise>

<exercise id='4' title='Test Your Knowledge'>

> You are writing a book using version control. In your local repository you have text file for each chapter. Today, after reading your work again and decided to move some of the text you have written for `chapter3.md` to `chapter1.md`. How would you save this change in Git?

Reviewer: ```suggestion
You are writing a book using version control. In your local repository, you have a text file for each chapter. Today, after reading your work again you decided to move some of the text you have written for `chapter3.md` to `chapter1.md`. How would you save this change in Git?
```

<choice>

> <opt text='It would cut and copy the text into the new file, add both files to the staging area, and then commit with a message explaining the changes.' correct='true'>

Reviewer: ```suggestion
<opt text='I would cut and copy the text into a new file, add both files to the staging area, and then commit with a message explaining the changes.' correct='true'>
```

</opt>

<opt text='I would <b>revert</b> to the commit I did after writing that paragraph. Then, I would add the text in the correct chapter and make a new commit with a message explaining the changes.'>

> Are you sure you want to loose all the changes you did for `chapter2.md`? Even if you can recover those changes, reverting the changes will make you go back to that particular point in the the history of your project.

Reviewer: ```suggestion
Are you sure you want to lose all the changes you made in `chapter2.md`? Even if you can recover those changes, reverting the changes will make you go back to that particular point in the the history of your project.
```

</opt>

<opt text='I would perform a <b>hard reset</b> to the commit I did after writing that paragraph. Then, I would add the text in the correct chapter and make a new commit with a message explaining the changes.'>

> When you realize a hard reset, you are deleting the history of your project until that commit. This means that you will be losing all the changes you did in between that commit.

Reviewer: ```suggestion
When you do a hard reset, you are deleting the history of your project up to a particular commit. This means that you will lose all the changes you made between that particular commit and your most recent one.
```

</opt>
</choice>

TRUE OR FALSE

> Select all the situations when it is a good idea to use reset your work to a previous commit

Reviewer: ```suggestion
Select all the situations in which it is a good idea to reset your work to a previous commit:
```

<choice>

> <opt text='When you are working in a project locally, without pushing to the remote' correct='true'>

Reviewer: ```suggestion
<opt text='When you are working on a project locally, without having already pushed your work to the remote' correct='true'>
```

If you are working locally on a private project, it is quite safe to use reset. Nobody is going to be affected by you deleting part of the history of your project.

</opt>
<opt text='When we are working with other colleagues pushing to the same repository' >

This is not a good idea because you are deleting part of the history of your project. Consider that your colleagues are working with the full history. This could be a problem at the moment of combining the information. In this case probably it is better to revert the changes, a command that we are going to learn in the next sections of this module.

</opt>

<opt text='Both are correct' >

> You can hard reset your work in both cases, but one option could generate problems.

Reviewer: ```suggestion
You can hard-reset your work in both cases, but one of the options could cause problems.
```

</opt>

</choice>
</exercise>

<exercise id='5' title='Revert your changes' type='slides, video'>

> <slides source='module4/module4_04' shot='0' start='3:42' end='4:35'> </slides>

Reviewer: I think this should be `module4/module4_03`; right now I see merge conflict slides when I click on the "Revert your changes" slide deck.

```suggestion
<slides source='module4/module4_03' shot='0' start='3:42' end='4:35'> </slides>
```

</exercise>

<exercise id='6' title='Test Your Knowledge'>

> What of this statements are TRUE about git revert

Reviewer: ```suggestion
Which of the following statements are TRUE about <code>git revert</code>?
```

<choice>
<opt text='This command creates a new commit that undoes the changes from a previous commit.' >
</opt>

> <opt text='This command undoes the changes of your project until the commit id that you determine.'>

Reviewer: ```suggestion
<opt text='This command undoes the changes in your project up to the commit id that you have specified.'>
```

</opt>
<opt text='This command makes shorter the history of the project (deletes commits)'>
</opt>
<opt text='This command is used to undo changes in your working directory that have not been comitted yet.'>
</opt>
<opt text='A and B are correct' correct='true'>
</opt>
<opt text='A and C are correct'>
</opt>
</choice>

</exercise>

<exercise id='7' title="Deal with merge conflicts"  type='slides, video'>

> <slides source='module4/module4_03' shot='0' start='3:42' end='4:35'> </slides>

Reviewer: ```suggestion
<slides source='module4/module4_04' shot='0' start='3:42' end='4:35'> </slides>
```

</exercise>

<exercise id='8' title='Test Your Knowledge'>

```
<<<<<<< HEAD
Copyright (c) 2022 Florencia D'Andrea
=======
Copyright (c) 2022 Name
>>>>>>> 3173605e73175bd86a3ee8cb078198040a3b3cf6
```

What do you have to do if you are working with the terminal and you want to write a different text for this merge conflict?

<choice>
<opt text='Accept current change' >
</opt>
<opt text='Accept incoming change'>
</opt>
<opt text='Accept both changes' correct='true'>
</opt>

> <opt text='You should delete everything between <code> <<<<<<< </code> and <code> >>>>>>> </code> add the text you want instead'>

Reviewer: ```suggestion
<opt text='You should delete everything between <code> <<<<<<< </code> and <code> >>>>>>> </code> and add the text you want instead'>
```
